### ISSUE TITLE
Spec gap: engine has no defined behaviour for self-dealt insurance absorption

### ISSUE THREAD

**Issue by abishekk92** [aeyakovenko/percolator#49]
# Spec gap: engine has no defined behaviour for self-dealt insurance absorption

## TL;DR (machine-readable)

- **Spec gap:** the engine's safety invariants (conservation, bilateral OI, haircut monotonicity) do not constrain the relationship between the owner of a bankrupt position and the owner of the position collecting the winning matured PnL. An attacker running two keypairs they both control can open opposing leveraged positions, induce a bankruptcy on one side, and extract value from the shared insurance pool through the winning side. No current engine invariant rules this out; the spec is silent.
- **Root cause** (restated): the engine tracks authorization per slot (pubkey), not per economic actor. `absorb_protocol_loss → use_insurance_buffer` (`src/percolator.rs:2288-2301`) absorbs the bankrupt leg's shortfall from shared insurance with no accounting for *who* the bankrupt owner is. This is a *completeness gap in the spec*, not a missing check in the code.
- **Empirical motivation:** drain is reachable on `aeyakovenko/percolator-prog` mainnet market `5ZamU…kTqB` right now (details in `aeyakovenko/percolator-prog#43`). Full 5 SOL insurance seed extractable at 2 SOL/leg attacker capital; linear scaling down to 0.01 SOL/leg with no economic floor. @alsk1992 and I independently confirmed the two-keypair bypass against `aeyakovenko/percolator-prog#37`.
- **Proposed invariant:** *for any owner `O`, the total amount absorbed from the insurance pool across all bankruptcies of positions owned by `O` must not exceed `O`'s cumulative attributable insurance contributions.* Plain language: each owner self-insures first; shared-pool access is bounded by skin-in-the-game.
- **Proposed implementation:** per-owner insurance contribution PDA, derived from `(market_pubkey, owner_pubkey)` — same pattern as SPL Associated Token Accounts. On bankruptcy, `absorb_protocol_loss` debits the bankrupt owner's PDA first; any residual shortfall falls through to the existing `record_uninsured_protocol_loss` path (which is already a no-op, leaving the current haircut mechanism to handle it). **Zero slab layout change.** Storage scales with active-owner count, paid by the owner who creates the PDA.
- **Why this closes the drain:** the attacker's two keypairs each derive a distinct PDA. Neither PDA can fund the other's contribution without real SOL outflow from the attacker's pocket. Self-dealt bankruptcy absorption is arithmetically bounded by the attacker's own prior contribution; net extraction from shared insurance is ≤ 0 by construction.
- **One open design question:** fee-attribution policy — which fee flows credit an owner's PDA? Three sub-options below; recommended default is a trading-fee share plus optional explicit top-up.
- **Reviewer ask:** (1) accept the spec-gap framing and the proposed invariant; (2) pick the fee-attribution policy; (3) I'm happy to author the implementation PR.

### The primitive

A self-dealt bankruptcy extracts value from shared insurance:

1. Entity opens opposing leveraged positions on both sides of a trade.
2. Oracle moves enough to bankrupt one side (loss > capital).
3. `LiquidateAtOracle` realises the loss at the new oracle price.
4. `absorb_protocol_loss` (`src/percolator.rs:2323`) → `use_insurance_buffer`
   (`src/percolator.rs:2291`) pays the bankruptcy shortfall from the
   insurance fund.
5. The winning side collects full matured PnL — residual grows in
   lockstep with the insurance absorption (@Copenhagen0x's "F7"
   analysis in `aeyakovenko/percolator-prog#39`).
6. Entity withdraws the winning side. Net extraction = insurance
   absorbed.

### Why wrapper-level checks don't close this

Both the proposed upstream fixes at the wrapper level —
`aeyakovenko/percolator-prog#37` (@Genmin) and
`aeyakovenko/percolator-prog#41` (this author) — implement
`owners_distinct_ok(lhs, rhs) = lhs != rhs` on trade paths. That's
a **syntactic** check: "two pubkeys differ." It closes the trivial
case where an attacker reuses a single `Keypair::new()`. It does not
close the case where the same human generates two keypairs. @alsk1992
independently identified and posted this bypass on
`aeyakovenko/percolator-prog#37`; I re-ran the drain PoCs against
both `aeyakovenko/percolator-prog#37`'s and
`aeyakovenko/percolator-prog#41`'s branches and confirmed full
extraction. The two-key cartel still drains the full insurance seed
on mainnet-scale parameters at 2 SOL per leg.

**The security property at stake — "one economic actor cannot be on
both sides of a trade that socialises losses" — is not expressible
in keypair equality.** On-chain identity is keypairs; economic
identity is controllers. Rather than trying to *detect* controllers
(which is unenforceable on-chain), the proposal below makes
controller identity *irrelevant*: each pubkey self-insures, and the
attacker's multi-keypair rotation gains no arbitrage over a single
trader acting in good faith.

### Candidate invariant

> **For any owner `O`:** the cumulative amount absorbed from the
> insurance pool across all bankruptcies of positions owned by `O`
> must not exceed `O`'s cumulative attributable insurance
> contributions.

Formally: `Σ absorbed_from_pool(bankruptcy_of(O)) ≤ Σ contributions_attributed_to(O)`.

One-liner, arithmetic, no controller-identity semantics. Enforceable
per-account via a simple accounting rule: *"absorb from `O`'s
contribution budget first; only fall through to the shared pool when
`O`'s budget is exhausted."*

### What existing spec guarantees (for comparison)

- **Conservation:** `vault >= c_tot + insurance`. Preserved under the
  drain — attacker value lands in matured PnL, which is "owed" to
  the attacker's winning leg. Conservation doesn't distinguish who's
  owed.
- **Bilateral OI:** positions net to zero. Preserved.
- **Haircut monotonicity:** residual-based haircut only tightens.
  Preserved.
- **Oracle monotonicity:** one-way clock on `publish_time`. Preserved.

The drain primitive violates none of these. That's exactly the spec
gap: the economic property we want is orthogonal to the structural
properties the spec currently names. The proposed invariant closes
the gap with a per-owner arithmetic constraint.

### Why this closes the drain

Attacker `A` opens two positions under keypairs `K1` and `K2`, both
with fresh PDAs (contribution = 0). One position bankrupts. Under
the tightened spec:

1. `absorb_protocol_loss(shortfall, bankrupt_owner=K1)` attempts to
   debit `K1`'s contribution PDA — **balance 0** — absorption from
   pool = `min(shortfall, 0) = 0`.
2. Full shortfall falls through to `record_uninsured_protocol_loss`
   (existing no-op).
3. Insurance fund `I` is unchanged. Residual `R = V - C_tot - I`
   stays at its pre-bankruptcy value.
4. `K2`'s matured PnL claim against residual is bounded by the
   existing haircut mechanism, which now correctly caps at
   residual's natural size (no insurance-absorption inflation).
5. Net extraction ≤ 0.

For honest traders `Alice` and `Bob`: identical behaviour **if
neither has built contribution balance yet**. Once either has
accumulated a contribution (via fee flow, see below), their own
bankruptcy draws from their own PDA first — which is the intuitive
"my fees went into insurance, so my shortfall gets covered first"
model. Honest case is never *worse* than today; in the steady state
where users have built contribution over time, honest coverage is
actually *better*.

### Account layout

One PDA per `(market, owner)` pair:

- **Seeds:** `[b"percolator_insurance", market_pubkey.as_ref(), owner_pubkey.as_ref()]`
- **Program:** percolator engine
- **Data (≤ 64 bytes):**
  ```rust
  #[repr(C)]
  pub struct InsuranceContribution {
      pub bump: u8,
      pub owner: [u8; 32],       // binding, matches seed
      pub market: [u8; 32],      // binding, matches seed
      pub balance: u128,         // cumulative contribution, lamport-scale
      pub _reserved: [u8; 8],    // for future fields
  }
  ```
- **Lifecycle:** lazy-initialised on first credit. Owner pays rent.
  Matches SPL Associated Token Account pattern exactly.

### Read / write sites

**Credit** (new contribution):
- On `Trade*` handlers, the portion of the trading fee that would
  currently flow to `insurance_fund.balance` is instead split
  `(trader_share, lp_share)` and credited to the trader's and LP's
  respective PDAs. Specifics in "Fee-attribution policy" below.
- On a new `TopUpInsuranceContribution { owner, amount }`
  instruction: explicit credit by any signer.

**Debit** (bankruptcy absorption):
- In `absorb_protocol_loss(loss, bankrupt_idx)`:
  ```rust
  let bankrupt_owner = self.accounts[bankrupt_idx].owner;
  let contribution_pda = derive_contribution_pda(market, bankrupt_owner);
  let absorbed = min(loss, contribution_pda.balance);
  contribution_pda.balance -= absorbed;
  let residual_loss = loss - absorbed;
  if residual_loss > 0 {
      self.record_uninsured_protocol_loss(residual_loss);  // existing no-op
  }
  ```
- Crucially, `insurance_fund.balance` (the aggregate field) either
  (a) disappears entirely, replaced by the sum over all PDAs, or
  (b) is reinterpreted as a shared-backstop pool (open design
  question — see below).

### Slab impact

**Zero.** The slab account layout is unchanged. The new PDA is a
separate account class. Kani proofs on slab state are unaffected;
the new invariant lives purely in PDA-state relations.

### Migration

- Pre-upgrade accounts: no PDA → balance 0 → bankruptcy absorption
  from pool is 0 → fall-through to existing residual mechanism.
- Post-upgrade: first trade or top-up creates the PDA, owner pays
  rent.
- Existing deployed markets (e.g. `5ZamU…kTqB`): post-upgrade
  drains are blocked because every bankrupt owner starts at
  contribution=0. **Strictly no worse than today** for the retroactive
  mainnet question; gets better as honest traders build contribution.

### Why this matches the ATA pattern

Every Solana developer recognises the `(program, owner, mint) → PDA`
derivation. Substituting `(percolator, owner, market)` yields a
directly-analogous on-chain representation, familiar authorization
model (owner signs to top up, engine CPI to debit), and migration
pattern (lazy init). Reviewing this design adds zero new concepts a
Solana dev must learn.

## Fee-attribution policy (one open design question)

The spec needs a definition of "attributable contribution." Three
sub-options, not mutually exclusive:

**F1. Trading-fee share.** Each `Trade*` op already charges a
`trading_fee_bps` fee on notional. Today the split is
`(lp_share, insurance_share, vault_share)`; the `insurance_share`
currently flows to `insurance_fund.balance`. Under the new model,
the insurance share is further split between the user side and the
LP side and credited to their respective PDAs. Fully passive — any
active trader accumulates contribution.

**F2. Explicit top-up.** A new permissionless `TopUpInsuranceContribution`
instruction lets any signer credit a specific owner's PDA. Explicit,
opt-in, useful for whitelisted market-makers who want to pre-fund
their coverage before trading.

**F3. Hybrid (recommended).** F1 as the default mechanism, with F2
available for explicit seeding. Covers both active traders (F1) and
operators who want to bootstrap coverage (F2).

The recommended default is **F3**. The only real design parameter
is the insurance-share ratio within the existing `trading_fee_bps` —
which is already a `config` field that markets set at init. No new
economic levers required.

### New invariant to prove

> For every trace reaching `absorb_protocol_loss`:
> `Σ (insurance_pool_decrements_attributed_to owner_O) ≤ Σ (credits_attributed_to owner_O)`

Proof obligation: every `absorb_protocol_loss` call site correctly
gates the debit on the bankrupt owner's PDA balance.

### Preservation of existing invariants

- **Conservation** (`vault >= c_tot + insurance_pool_total`):
  unchanged; `insurance_pool_total` is now reinterpreted as either
  the sum of PDAs or a separate shared-backstop PDA (design question
  above). Conservation holds per either interpretation.
- **Bilateral OI:** unchanged. PDA state is orthogonal to position
  accounting.
- **Haircut monotonicity:** unchanged. Residual formula is
  unaffected because `I` doesn't inflate from self-dealt bankruptcy
  absorptions anymore.

### Fuzz suite extensions

- Property: for every owner, `Σ debits_from_PDA ≤ Σ credits_to_PDA`.
- Property: for every self-dealt bankruptcy pattern (enumerable at
  small state space), attacker net extraction ≤ 0.
- Property: honest trader PDA balance monotonically non-decreasing
  under trading activity alone (no absorption events).

## Interaction with existing filings

- **`aeyakovenko/percolator-prog#41`** — wrapper-level hardening
  (reject `min_oracle_price_cap_e2bps = 0` at init, extend admin-burn
  gate, same-owner check on trades). Scoped to wrapper; closes the
  single-keypair drain subclass at the wrapper level; spec-independent.
  Merges regardless of this memo.
- **`aeyakovenko/percolator-prog#43`** — consolidated findings ledger
  for wrapper-side drain surface. References this memo as the spec
  discussion for the two-key cartel row of the attack matrix.
- **`aeyakovenko/percolator-prog#37`** — @Genmin's wrapper-level
  `owners_distinct_ok` check. Same single-keypair close as
  `aeyakovenko/percolator-prog#41`'s first change; bypassed by the
  same two-key cartel this memo addresses.
- **`aeyakovenko/percolator-prog#39`** — @Copenhagen0x's F7
  vault-debit proposal. Equivalent to an unconditional winner-side
  haircut, which blocks the drain but causes the documented
  honest-case regression at `src/percolator.rs:2311-2317`. This
  proposal avoids that regression entirely by never touching the
  winner side — only the loser's draw from the pool is constrained,
  and only when the loser lacks own-PDA coverage.

## Alternatives considered (and why rejected)

Briefly, for intellectual honesty:

- **Counterparty stamp on `Account`.** Each account stores the most
  recent (or a set of) trade counterparty owners; `absorb_protocol_loss`
  checks for self-dealt pairs. *Rejected because:* requires defining
  "common controller" at the spec level with no canonical answer
  (same-owner? same init-depositor? K-set intersection?), all
  variants have bypass paths, and the slab layout grows by
  `32 B × MAX_ACCOUNTS`.
- **Unconditional winner-side haircut scaling** (@Copenhagen0x's
  `aeyakovenko/percolator-prog#39`). *Rejected because:* honest-case
  regression documented at `src/percolator.rs:2311-2317`.
- **Permissioned LP registration.** *Rejected because:* anti-
  permissionless; doesn't address user-vs-user variants; whitelist
  becomes a new governance surface.

## Reviewer ask

1. **Accept** the spec-gap framing and the proposed invariant
   (*"own absorption bounded by own contribution"*).
2. **Pick** the fee-attribution policy (F1 / F2 / F3; recommended F3).
3. **Confirm** the shared-backstop-pool vs. fully-private
   decomposition (I'd lean fully-private; it's the cleanest match
   for the invariant, and the shared pool's drain-exposure gap under
   a backstop model would need a second-order mechanism to close).
4. If all three land, I'm happy to author the implementation PR
   (PDA type, seeds, trade-handler fee-attribution changes,
   `absorb_protocol_loss` rewrite, Kani invariant, fuzz extensions).
